docs/src/parameters.md
# Workflow Parameters

!!! note
    Due to some cromwell parsing issues, all options and arguments must be provided as strings unless stated otherwise. That is, in your input file you should write `npcs="10"`, not `npcs=10`.

## Required Arguments

These arguments must be provided or the workflow will not run.

- `covariates_file`: A CSV file containing the set of covariates and phenotypes. Missing values can be either NAN, NA, NULL or an empty character.
- `genotypes`: A PLINK BED fileset in GRCh38.
- `imputed_genotypes`: A set of PLINK PGEN filesets, one for each chromosome in GRCh38.
- `phenotypes`: A list of phenotypes for which each GWAS will be run independently.

## General Options

These options describe the general behaviour of the workflow.

- `groupby` (default: []): A list of variables used to stratify individuals. If empty, the full dataset is used.
- `filterby` (default: []): A list of conditions on covariates present in the dataset to filter individuals. At the moment this can either be in the format:
  - `COLUMN=value`: which can be used to filter for individuals with categorical traits, e.g., `SUPERPOPULATION=AFR` if the `covariates_file` has a `SUPERPOPULATION` column.
  - `COLUMN>value`: which can be used to filter individuals with continuous variables, e.g., `AGE>70`. Any of `<`, `>`, `=`, `<=`, `>=` is a valid operator.
  - A combination of the above, e.g., `filterby=[SUPERPOPULATION=AFR, AGE>70]`.
- `covariates` (default: `[AGE, SEX, AGE_x_AGE, AGE_x_SEX]`): A list of covariates used to adjust for confounding or increase power in the association testing step. Product of variables can be defined using the `_x_` syntax, for example: `AGE_x_SEX`.
- `min_cases_controls` (default: 10): For binary traits only, the minimum number of cases/controls within a group to proceed to the GWAS.

## PCA Options

PCA is performed using [plink2's PCA](https://www.cog-genomics.org/plink/2.0/strat).

- `npcs` (default 10): Number of principal components to use to account for population structure.
- `ip_values` (default: `1000 50 0.05`): A string of values used to create independent genotypes for PCA (see [here](https://www.cog-genomics.org/plink/2.0/ld)).
- `approx_pca` (default: true): Whether to use an approximation to the PCA algorithm (see [here](https://www.cog-genomics.org/plink/2.0/strat)). Turning this to `false` if used only for small datasets, for instance during testing.

## GWAS Options

Genome-wide association testing is performed using [REGENIE](https://rgcgithub.github.io/regenie/options/).

- `regenie_cv_folds` (default: 5): Number of folds for Regenie step 1. Can also be `loocv`.
- `regenie_bsize` (default 1000): Regenie block size.

## Fine-Mapping Options

Finemapping proceeds in two stages. First clumps are formed using [plink2 LD-based result clumping](https://www.cog-genomics.org/plink/2.0/postproc) and lead variants are identified. Then a window is formed around the lead variant to be further finemapped with [SuSiE](https://stephenslab.github.io/susieR/).

- `Xtype` (default: `dosages`): To use `dosages` or `genotypes` during finemapping.
- `min_sig_clump_size` (default: 3): Defines the minimum number of variants within a clump for a locus to be considered for finemapping.
`lead_pvalue` (default: `5e-8`): A clump's lead variant must have at least this p-value.
- `p2_pvalue` (default: 1e-5): Other variants in the clump must have at least this p-value.
- `r2_threshold` (default: 0.1): The LD between the lead variant and other variants in the clump must also as strong as `r2_threshold`.
- `clump_kb` (default: 500): The maximum distance between the lead variant and a clumped variant.
- `n_causal` (default: 10): SuSiE's `L` variable, the expected number of causal variants in a locus.
- `finemap_window_kb` (default: 500): Defines the size of the finemapping window.

## Meta-Analysis Options

Meta analysis is performed using [METAL](https://github.com/statgen/METAL).

- `meta_analysis` (default: `length(groupby) > 0`): The only parameter that must be provided as a boolean (not a string). It defaults to true if groups are provided and false otherwise.
- `meta_exclude` (default: ["ADMIXED"]): Whether some groups should be excluded from the meta-analysis stage. Note that this will result in lower power. We default to exclusing admixed individuals if the grouping column contains superpopulations.
- `meta_method` (default: "STDERR"). The METAL's meta analysis method, please refer to their documentation for more information.

## Miscellaneous Options

- `maf` (default: 0.01): Minor allele frequency threshold. This is used to filter variants for PCA and for plotting GWAS results. The association testing step is always performed across all variants. Post fitering for downstream analysis is left to the user.
- `mac` (default: 10): Minor allele count used to filter variants in REGENIE and PCA QC.

## Developpement Options

In principle you only need to change these if you are trying to contribute to the workflow.

- `docker_image` (default: `olivierlabayle/wdl-gwas:main`): The docker image used to run the workflow. You shouldn't need to change that if you are not contributing to the workflow.
- `julia_use_sysimage` (default: `true`): Whether to use the Julia's system image built within the container with [PackageCompiler](https://julialang.github.io/PackageCompiler.jl/stable/). During development it is better to have this to false to reflect the current state of the code.
- `julia_threads` (default: `auto`): Julia's number of [threads](https://docs.julialang.org/en/v1/manual/multi-threading/). Some external tools seem to crash the parent process when multiple threading is used locally. Typically set to false in the development stage.



src/cli.jl
function cli_settings()
    s = ArgParseSettings(
        description="WDL-GWAS CLI",
        add_version = true,
        commands_are_required = false,
        version=string(pkgversion(PopGen))
    )

    @add_arg_table! s begin
        "meta-analyse"
            action = :command
            help = "Runs meta-analysis across GWAS results."

        "finemap"
            action = :command
            help = "Runs fine-mapping analysis."

        "make-plots"
            action = :command
            help = "Generates GWAS plots."

        "pca-qc"
            action = :command
            help = "Runs PCAb-ased QC on genotypes to exclude outlier variants."

        "make-groups-and-covariates"
            action = :command
            help = "Generates groups, phenotypes and covariates files."
        
        "merge-covariates-pcs"
            action = :command
            help = "Merges covariates and PCs files."

        "merge-chr-results"
            action = :command
            help = "Merges REGENIE results from different chromosomes."
    end

    @add_arg_table! s["meta-analyse"] begin
        "gwas-results-list"
            arg_type = String
            required = true
            help = "List of REGENIE results files to be meta-analysed."
    
        "--exclude"
            arg_type = String
            help = "List of groups to exclude from the meta-analysis, comma separated."

        "--method"
            arg_type = String
            help = "Meta-analysis method to use (fixed or random)."
            default = "STDERR"

        "--output-prefix"
            arg_type = String
            help = "Prefix to output files."
            default = "gwas.meta_analysis"
    end

    @add_arg_table! s["finemap"] begin
        "gwas-results-file"
            arg_type = String
            required = true
            help = "Path to the GWAS results file."
        "pgen-prefix"
            arg_type = String
            required = true
            help = "Prefix for the PGEN fileset (without .pgen extension)."
        "covariates-file"
            arg_type = String
            required = true
            help = "Path to the covariates file (TSV format)."
        "sample-file"
            arg_type = String
            required = true
            help = "Path to the sample IDs file used to generate the GWAS results."
        "--Xtype"
            arg_type = String
            help = "Type of genotype data to use for fine-mapping, either `:dosages` or `:genotypes`."
            default = "dosages"
        "--output-prefix"
            arg_type = String
            help = "Prefix to output the significant clumps (TSV format)."
            default = "finemapping_results"
        "--min-sig-clump-size"
            arg_type = Int
            help = "Minimum number of variants in a clump to be considered significant."
            default = 3
        "--lead-pvalue"
            arg_type = Float64
            help = "P-value threshold for lead variants in clump."
            default = 5e-8
        "--p2-pvalue"
            arg_type = Float64
            help = "P-value threshold for secondary variants in clump."
            default = 1e-5
        "--r2-threshold"
            arg_type = Float64
            help = "R2 threshold to consider variants in LD for clumping."
            default = 0.5
        "--clump-kb"
            arg_type = Int
            help = "Window size (in kb) to consider variants in LD for clumping."
            default = 250
        "--n-causal"
            arg_type = Int
            help = "Number of causal variants to assume in fine-mapping."
            default = 10
        "--finemap-window-kb"
            arg_type = Int
            help = "Window size (in kb) to compute LD matrix for fine-mapping."
            default = 1000
        "--phenotype"
            arg_type = String
            help = "Name of the phenotype column in the covariates file."
            default = "Y"
        "--rss"
            help = "Whether to run fine-mapping using summary statistics (RSS)"
            action = :store_true
        "--exclude"
            arg_type = String
            help = "List of groups to exclude from the sample file, comma separated."
            default = ""
    end

    @add_arg_table! s["merge-chr-results"] begin
        "merge-list"
            arg_type = String
            required = true
            help = "File containing list of files to be merged."
        "--output-prefix"
            arg_type = String
            help = "Output path"
            default = "results.all_chr"
    end

    @add_arg_table! s["make-plots"] begin
        "gwas-results"
            arg_type = String
            required = true
            help = "Path to GWAS results file."

        "finemapping-results"
            arg_type = String
            required = true
            help = "Path to finemapping results file."

        "--maf"
            arg_type = Float64
            help = "Minor allele frequency threshold to filter results."
            default = 0.01
        
        "--output-prefix"
            arg_type = String
            help = "Prefix to output files."
            default = "gwas.plot"
    end

    @add_arg_table! s["merge-covariates-pcs"] begin
        "covariates-file"
            arg_type = String
            required = true
            help = "Path to covariates file."
        
        "pcs-prefix"
            arg_type = String
            required = true
            help = "Prefix to PCs files."

        "--output"
            arg_type = String
            help = "Output file name."
            default = "covariates_and_pcs.csv"
    end

    @add_arg_table! s["make-groups-and-covariates"] begin
        "covariates-file"
            arg_type = String
            required = true
            help = "Path to covariates file."

        "--groupby"
            arg_type = String
            help = "Comma separated list of variables to use to stratify the GWAS."
            default = nothing

        "--filters"
            arg_type = String
            help = "Filters to apply to the data."
            default = nothing

        "--covariates"
            arg_type = String
            help = "Comma separated list of covariates to include in the output file."
            default = "AGE"

        "--phenotypes"
            arg_type = String
            help = "Comma separated list of phenotypes to include in the output file."
            default = "SEVERE_COVID_19"

        "--output-prefix"
            arg_type = String
            help = "Prefix to output files."
            default = "group"
        
        "--min-cases-controls"
            arg_type = Int
            help = "Minimum group size."
            default = 100
    end

    return s
end

function julia_main()::Cint
    settings = parse_args(ARGS, cli_settings())
    cmd = settings["%COMMAND%"]
    @info "Running WDL-GWAS CLI: $cmd"
    cmd_settings = settings[cmd]
    if cmd == "make-groups-and-covariates"
        make_groups_and_covariates(
            cmd_settings["covariates-file"];
            groupby_string=cmd_settings["groupby"],
            covariates_string=cmd_settings["covariates"],
            phenotypes_string=cmd_settings["phenotypes"],
            output_prefix=cmd_settings["output-prefix"],
            min_cases_controls=cmd_settings["min-cases-controls"],
            filters_string=cmd_settings["filters"]
        )
    elseif cmd == "merge-covariates-pcs"
        merge_covariates_and_pcs(
            cmd_settings["covariates-file"],
            cmd_settings["pcs-prefix"];
            output=cmd_settings["output"]
        )
    elseif cmd == "make-plots"
        make_plots(
            cmd_settings["gwas-results"],
            cmd_settings["finemapping-results"];
            maf=cmd_settings["maf"],
            output_prefix=cmd_settings["output-prefix"]
        )
    elseif cmd == "merge-chr-results"
        merge_chr_results(
            cmd_settings["merge-list"];
            output_prefix=cmd_settings["output-prefix"]
        )
    elseif cmd == "finemap"
        finemap_significant_regions(
            cmd_settings["gwas-results-file"],
            cmd_settings["pgen-prefix"],
            cmd_settings["covariates-file"],
            cmd_settings["sample-file"];
            Xtype=cmd_settings["Xtype"],
            output_prefix=cmd_settings["output-prefix"],
            min_sig_clump_size=cmd_settings["min-sig-clump-size"],
            lead_pvalue=cmd_settings["lead-pvalue"],
            p2_pvalue=cmd_settings["p2-pvalue"],
            r2_threshold=cmd_settings["r2-threshold"],
            clump_kb=cmd_settings["clump-kb"],
            n_causal=cmd_settings["n-causal"],
            finemap_window_kb=cmd_settings["finemap-window-kb"],
            phenotype=cmd_settings["phenotype"],
            rss=cmd_settings["rss"],
            exclude_string=cmd_settings["exclude"]
        )
    elseif cmd == "meta-analyse"
        meta_analyse(
            cmd_settings["gwas-results-list"];
            output_prefix=cmd_settings["output-prefix"],
            exclude_string=cmd_settings["exclude"],
            method=cmd_settings["method"],
        )
    else
        throw(ArgumentError(string("Unknown command: ", cmd)))
    end
    return 0
end